src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -162,6 +162,14 @@ def version(self):
             self._version = self.dist.parsed_version
         return self._version
 
+    def format_for_error(self):
+        # type: () -> str
+        return "{} {} (from {})".format(
+            self.name,
+            self.version,
+            self.link.file_path if self.link.is_file else self.link
```

Comment:
How does this look for cached paths?

Reply:
I'll try it tomorrow, but from my recollection "annoyingly long" ;-)

Reply:
Can we use the original link w/ `(cached)` in the name in that case?

OK to do this in a follow up, if you file an issue to aggregate all the call sites that print `req.link`.